File: backend/src/scheduler/core/arms.ts
import type { SchedulingArm } from "@zenflow/shared";
import { utcToMinutes } from "../../common/utils";

/**
 * The five Disjoint-LinUCB time-of-day bands
 * (`docs/adr/0001-linucb-model-design.md` §2). Ranges are minute-of-day in the
 * user's LOCAL wall clock, **half-open, lower-inclusive** — a slot starting
 * exactly at 17:00 (minute 1020) is `EVENING`.
 *
 * Pure: no I/O, no clock, no randomness.
 */
export interface ArmBand {
  arm: SchedulingArm;
  /** Minute of day, inclusive. */
  start: number;
  /** Minute of day, exclusive. */
  end: number;
}

export const ARM_BANDS: readonly ArmBand[] = [
  { arm: "EARLY_MORNING", start: 0, end: 360 },
  { arm: "MORNING", start: 360, end: 660 },
  { arm: "AFTERNOON", start: 660, end: 1020 },
  { arm: "EVENING", start: 1020, end: 1200 },
  { arm: "NIGHT", start: 1200, end: 1440 },
];

/** The arm whose half-open band contains `minuteOfDay` (wrapped into `[0, 1440)`). */
export function armOfMinute(minuteOfDay: number): SchedulingArm {
  const m = ((Math.floor(minuteOfDay) % 1440) + 1440) % 1440;
  for (const band of ARM_BANDS) {
    if (m >= band.start && m < band.end) return band.arm;
  }
  // Unreachable — the bands tile [0, 1440) with no gap.
  return "NIGHT";
}

const MS_PER_MIN = 60_000;
const MINUTES_PER_DAY = 1440;

/**
 * Fraction of `[startMs, endMs)` whose local wall-clock minute-of-day falls
 * inside `arm`'s band, in `timezone`.
 *
 * A slot may span local midnight (D5): it is split at each local-midnight
 * boundary and the same-day overlap of every segment with the band is summed.
 * The band identifiers repeat each day, so e.g. a 23:00–01:00 slot contributes
 * to `NIGHT` (23:00–24:00) and `EARLY_MORNING` (00:00–01:00).
 */
export function overlapRate(
  startMs: number,
  endMs: number,
  arm: SchedulingArm,
  timezone: string,
): number {
  if (endMs <= startMs) return 0;

  const band = ARM_BANDS.find((b) => b.arm === arm);
  if (!band) throw new Error(`overlapRate: unknown arm ${arm}`);

  const totalMinutes = (endMs - startMs) / MS_PER_MIN;
  let overlapMinutes = 0;

  let cursor = startMs;
  while (cursor < endMs) {
    const segStartMin = utcToMinutes(new Date(cursor), timezone);
    // End of the local day containing `cursor`.
    const dayEndMs = cursor + (MINUTES_PER_DAY - segStartMin) * MS_PER_MIN;
    const segEndMs = Math.min(dayEndMs, endMs);
    const segEndMin = segStartMin + (segEndMs - cursor) / MS_PER_MIN;

    overlapMinutes += Math.max(
      0,
      Math.min(segEndMin, band.end) - Math.max(segStartMin, band.start),
    );
    cursor = segEndMs;
  }

  return overlapMinutes / totalMinutes;
}

File: services/bandit/src/core/arms.py
"""LinUCB time-of-day arm bands (port of ``arms.ts``)."""

from __future__ import annotations

import math

from .slot import utc_to_minutes

ARM_BANDS: tuple[tuple[str, int, int], ...] = (
    ("EARLY_MORNING", 0, 360),
    ("MORNING", 360, 660),
    ("AFTERNOON", 660, 1020),
    ("EVENING", 1020, 1200),
    ("NIGHT", 1200, 1440),
)


def arm_of_minute(minute_of_day: float) -> str:
    m = math.floor(minute_of_day) % 1440
    for arm, start, end in ARM_BANDS:
        if start <= m < end:
            return arm
    return "NIGHT"


def overlap_rate(start_ms: int, end_ms: int, arm: str, timezone: str) -> float:
    """Fraction of ``[start, end)`` whose local minute-of-day is inside ``arm``."""
    if end_ms <= start_ms:
        return 0.0
    band = next((b for b in ARM_BANDS if b[0] == arm), None)
    if band is None:
        raise ValueError(f"overlap_rate: unknown arm {arm}")
    _, b_start, b_end = band
    total = (end_ms - start_ms) / 60_000
    overlap = 0.0
    cursor = start_ms
    while cursor < end_ms:
        seg_start_min = utc_to_minutes(cursor, timezone)
        day_end = cursor + (1440 - seg_start_min) * 60_000
        seg_end = min(day_end, end_ms)
        seg_end_min = seg_start_min + (seg_end - cursor) / 60_000
        overlap += max(0.0, min(seg_end_min, b_end) - max(seg_start_min, b_start))
        cursor = seg_end
    return overlap / total
